File: java/src/test/java/io/github/muthuishere/toolnexus/RagConsumerTest.java
package io.github.muthuishere.toolnexus;

import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.ProxySelector;
import java.net.SocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CancellationException;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Tests for the ADR-0001 rag_go consumer needs (7 gaps) + the DisableTools/DisableSkills
 * ergonomic layer, ported to the Java port. Mirrors {@code golang/rag_consumer_test.go}.
 * Client-side gaps use an in-process {@link HttpServer} that captures the decoded request body;
 * MCP-side gaps stand up a real streamable-HTTP MCP server by serving a toolkit (the same pattern
 * {@link McpServeTest} uses, since the Java MCP SDK ships no in-memory transport). Hermetic — no
 * network, no live LLM.
 */
class RagConsumerTest {

    // Response body that satisfies BOTH the openai (choices) and anthropic (content) decoders.
    private static final String LLM_RESPONSE =
            "{\"choices\":[{\"message\":{\"content\":\"ok\"}}],"
                    + "\"content\":[{\"type\":\"text\",\"text\":\"ok\"}],"
                    + "\"usage\":{\"prompt_tokens\":1,\"completion_tokens\":1,\"total_tokens\":2,"
                    + "\"input_tokens\":1,\"output_tokens\":1}}";

    /** An ephemeral LLM server that records the last decoded request body. */
    private static final class CaptureLLM implements AutoCloseable {
        final HttpServer server;
        final AtomicReference<Map<String, Object>> last = new AtomicReference<>();

        CaptureLLM() throws IOException {
            server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
            server.createContext("/", ex -> {
                byte[] raw = ex.getRequestBody().readAllBytes();
                try {
                    last.set(Json.toMap(new String(raw, StandardCharsets.UTF_8)));
                } catch (RuntimeException ignored) {
                    last.set(new LinkedHashMap<>());
                }
                byte[] b = LLM_RESPONSE.getBytes(StandardCharsets.UTF_8);
                ex.getResponseHeaders().add("Content-Type", "application/json");
                ex.sendResponseHeaders(200, b.length);
                try (OutputStream os = ex.getResponseBody()) {
                    os.write(b);
                }
            });
            server.start();
        }

        String url() {
            return "http://127.0.0.1:" + server.getAddress().getPort();
        }

        Map<String, Object> body() {
            return last.get();
        }

        @Override
        public void close() {
            server.stop(0);
        }
    }

    private static Map<String, Object> objSchema() {
        Map<String, Object> s = new LinkedHashMap<>();
        s.put("type", "object");
        s.put("properties", new LinkedHashMap<>());
        return s;
    }

    private static Tool mkTool(String name) {
        return NativeTool.of(name, name, objSchema(), args -> name);
    }

    /** Build a config map from alternating key/value pairs. */
    private static Map<String, Object> m(Object... kv) {
        Map<String, Object> map = new LinkedHashMap<>();
        for (int i = 0; i < kv.length; i += 2) map.put((String) kv[i], kv[i + 1]);
        return map;
    }

    private static List<String> names(List<Tool> ts) {
        List<String> out = new ArrayList<>();
        for (Tool t : ts) out.add(t.name());
        return out;
    }

    // ---- Gap 1: requestParams merge (caller wins) + bodyTransform last + forbidden keys ----

    @Test
    void gap1RequestParamsMergeAndCallerWins() throws IOException {
        try (CaptureLLM llm = new CaptureLLM()) {
            Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false));

            LlmClient oa = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("openai").model("m").apiKey("k")
                    .requestParams(m("temperature", 0.42, "chat_template_kwargs", m("enable_thinking", false))));
            oa.run("hi", tk);
            assertEquals(0.42, ((Number) llm.body().get("temperature")).doubleValue(), 1e-9);
            assertNotNull(llm.body().get("chat_template_kwargs"), "chat_template_kwargs missing");

            LlmClient an = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("anthropic").model("m").apiKey("k")
                    .requestParams(m("max_tokens", 999)));
            an.run("hi", tk);
            assertEquals(999, ((Number) llm.body().get("max_tokens")).intValue(),
                    "caller wins over the anthropic default 4096");

            tk.close();
        }
    }

    @Test
    @SuppressWarnings("unchecked")
    void gap1BodyTransformAndForbiddenKeys() throws IOException {
        try (CaptureLLM llm = new CaptureLLM()) {
            Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false));
            LlmClient c = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("openai").model("m").apiKey("k")
                    .requestParams(m("temperature", 0.5, "messages", List.of("nope"),
                            "tools", List.of(1), "stream", true))
                    .bodyTransform(b -> {
                        b.put("injected", "yes");
                        b.remove("temperature");
                        return b;
                    }));
            c.run("hi", tk);

            assertEquals("yes", llm.body().get("injected"), "bodyTransform output not sent");
            assertFalse(llm.body().containsKey("temperature"),
                    "bodyTransform ran before marshal but key survived");
            assertFalse(llm.body().containsKey("stream"),
                    "forbidden 'stream' override leaked into a non-stream call");
            List<Object> msgs = (List<Object>) llm.body().get("messages");
            assertNotNull(msgs);
            assertFalse(msgs.isEmpty(), "forbidden 'messages' override replaced the real messages");
            assertTrue(msgs.get(0) instanceof Map, "real messages should be role/content objects, not 'nope'");

            tk.close();
        }
    }

    // ---- Gap 2: injectable HTTP client (LLM path only) ----

    @Test
    void gap2HttpClientInjected() throws IOException {
        try (CaptureLLM llm = new CaptureLLM()) {
            AtomicInteger hits = new AtomicInteger(0);
            // A recording ProxySelector proves THIS client made the call; NO_PROXY keeps it direct.
            ProxySelector recording = new ProxySelector() {
                @Override
                public List<Proxy> select(URI uri) {
                    hits.incrementAndGet();
                    return List.of(Proxy.NO_PROXY);
                }

                @Override
                public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
                }
            };
            HttpClient injected = HttpClient.newBuilder().proxy(recording).build();

            Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false));
            LlmClient c = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("openai").model("m").apiKey("k")
                    .httpClient(injected));
            c.run("hi", tk);
            assertTrue(hits.get() > 0, "injected HttpClient was not used");
            tk.close();
        }
    }

    // ---- Gap 5: empty tool list omits tools/tool_choice; non-empty keeps them ----

    @Test
    void gap5OmitEmptyTools() throws IOException {
        try (CaptureLLM llm = new CaptureLLM()) {
            Toolkit empty = Toolkit.create(new Toolkit.Options().builtins(false));
            LlmClient c1 = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("openai").model("m").apiKey("k"));
            c1.run("hi", empty);
            assertFalse(llm.body().containsKey("tools"), "empty toolkit still sent tools");
            assertFalse(llm.body().containsKey("tool_choice"), "empty toolkit still sent tool_choice");
            empty.close();

            Toolkit one = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(mkTool("t")));
            LlmClient c2 = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("openai").model("m").apiKey("k"));
            c2.run("hi", one);
            assertTrue(llm.body().containsKey("tools"), "non-empty toolkit dropped tools");
            assertEquals("auto", llm.body().get("tool_choice"), "non-empty toolkit dropped tool_choice");
            one.close();
        }
    }

    // ---- Gap 4: conversationStore accessor (identity + default reflects saved transcript) ----

    @Test
    void gap4ConversationStoreAccessor() throws IOException {
        LlmClient.InMemoryConversationStore custom = new LlmClient.InMemoryConversationStore();
        LlmClient c1 = LlmClient.create(new LlmClient.Options()
                .baseUrl("http://x").style("openai").model("m").store(custom));
        assertTrue(c1.conversationStore() == custom, "accessor did not return the supplied store");

        try (CaptureLLM llm = new CaptureLLM()) {
            Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false));
            LlmClient c2 = LlmClient.create(new LlmClient.Options()
                    .baseUrl(llm.url()).style("openai").model("m").apiKey("k"));
            assertNotNull(c2.conversationStore(), "default store is nil");
            c2.ask("hi", tk, "conv1");
            List<Object> hist = c2.conversationStore().get("conv1");
            assertNotNull(hist);
            assertTrue(hist.size() >= 2,
                    "default store did not hold the saved transcript, got " + hist.size() + " msgs");
            tk.close();
        }
    }

    // ---- MCP-server helper: serve a toolkit exposing tools a,b,c as streamable-HTTP MCP ----

    private static A2AServer.ServeHandle startMCP() {
        Toolkit serverTk = Toolkit.create(new Toolkit.Options().builtins(false)
                .extraTools(mkTool("a"), mkTool("b"), mkTool("c")));
        return serverTk.serve("127.0.0.1:0", new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig()));
    }

    // ---- Gap 7: per-server tool allowlist (keyed on ORIGINAL name) ----

    @Test
    void gap7PerServerToolAllowlist() {
        A2AServer.ServeHandle srv = startMCP();
        String url = srv.url() + "/mcp";
        try {
            McpSource allow = McpSource.load(m("srv", m("url", url, "tools", m("a", true, "b", true))));
            List<String> allowNames = names(allow.tools());
            assertEquals(2, allowNames.size(), "allowlist tools = " + allowNames);
            assertTrue(allowNames.contains("srv_a") && allowNames.contains("srv_b"),
                    "allowlist tools = " + allowNames + ", want srv_a,srv_b");
            allow.close();

            McpSource drop = McpSource.load(m("srv", m("url", url, "tools", m("c", false))));
            List<String> dropNames = names(drop.tools());
            assertEquals(2, dropNames.size(), "drop-list tools = " + dropNames);
            assertFalse(dropNames.contains("srv_c"), "drop-list kept srv_c: " + dropNames);
            drop.close();

            McpSource all = McpSource.load(m("srv", m("url", url)));
            assertEquals(3, names(all.tools()).size(), "nil filter should load all 3");
            all.close();
        } finally {
            srv.stop();
        }
    }

    // ---- Gap 6: list-only inventory — UNFILTERED, with per-server status incl. a failed server ----

    @Test
    void gap6ListMcpTools() {
        A2AServer.ServeHandle srv = startMCP();
        String url = srv.url() + "/mcp";
        try {
            McpSource.McpInventory inv = McpSource.listMcpTools(m(
                    "good", m("url", url, "tools", m("a", true)),   // filter IGNORED by the inventory
                    "bad", m("url", "http://127.0.0.1:1/mcp", "timeout", 500)));
            assertEquals(3, inv.tools().get("good").size(),
                    "inventory should be unfiltered (a,b,c)");
            assertEquals("connected", inv.status().get("good"));
            assertEquals("failed", inv.status().get("bad"), "bad server should be failed");
            // original, unprefixed names carried through.
            List<String> good = new ArrayList<>();
            for (McpSource.ToolInfo ti : inv.tools().get("good")) good.add(ti.name());
            assertTrue(good.contains("a") && good.contains("b") && good.contains("c"),
                    "inventory names should be original/unprefixed: " + good);
        } finally {
            srv.stop();
        }
    }

    // ---- Gap 3: hanging server bounded by timeout → failed; cancel aborts the whole load ----

    @Test
    void gap3HangingBoundedAndCancel() throws IOException {
        // A server that accepts but never responds → connect/init bounded by the server timeout.
        HttpServer hang = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        CountDownLatch unblock = new CountDownLatch(1);
        hang.createContext("/", ex -> {
            try {
                unblock.await();
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
            }
        });
        hang.start();
        String hurl = "http://127.0.0.1:" + hang.getAddress().getPort() + "/mcp";
        try {
            long t0 = System.currentTimeMillis();
            McpSource src = McpSource.load(m("hang", m("url", hurl, "timeout", 500)));
            assertEquals("failed", src.status().get("hang"),
                    "hanging server should be bounded to failed by its timeout");
            assertTrue(System.currentTimeMillis() - t0 < 5000, "not bounded by timeout");
            src.close();

            // An already-signalled cancel aborts the whole load promptly with CancellationException.
            McpSource.CancelSignal cancelled = () -> true;
            long t1 = System.currentTimeMillis();
            assertThrows(CancellationException.class, () ->
                    McpSource.load(m("hang", m("url", hurl, "timeout", 60000)), null, cancelled));
            assertTrue(System.currentTimeMillis() - t1 < 5000, "cancel should abort promptly");
        } finally {
            unblock.countDown();
            hang.stop(0);
        }
    }

    // ---- DisableTools / DisableSkills ----

    @Test
    void disableToolsAndSkills() {
        A2AServer.ServeHandle srv = startMCP();
        String url = srv.url() + "/mcp";
        try {
            Toolkit tk = Toolkit.create(new Toolkit.Options()
                    .builtins(false)
                    .mcpConfig(m("srv", m("url", url)))
                    .disableTools("srv_b"));
            assertNull(tk.get("srv_b"), "DisableTools did not drop srv_b");
            assertNotNull(tk.get("srv_a"), "DisableTools dropped too much (srv_a gone)");
            tk.close();
        } finally {
            srv.stop();
        }

        Toolkit sk = Toolkit.create(new Toolkit.Options()
                .builtins(false)
                .skills(List.of(new SkillSource.SkillDef("keep", "k", "k"),
                        new SkillSource.SkillDef("drop", "d", "d")))
                .disableSkills("drop"));
        String prompt = sk.skillsPrompt();
        assertTrue(prompt.contains("keep"), "DisableSkills dropped the kept skill; prompt:\n" + prompt);
        assertFalse(prompt.contains("**drop**"), "DisableSkills wrong; prompt:\n" + prompt);
        sk.close();
    }
}

File: golang/rag_consumer_test.go
// Tests for the ADR-0001 rag_go consumer needs (7 gaps) + the DisableTools/
// DisableSkills ergonomic layer. Client-side gaps use an httptest LLM that
// captures the request body; MCP-side gaps use a real streamable-HTTP MCP server
// built from buildMcpServer (hermetic, no external network).
package toolnexus

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/mark3labs/mcp-go/server"
)

// TestMain re-execs this test binary as a real stdio MCP server (exposing a,b,c) when
// TN_STDIO_MCP=1, so TestStdioOutboundChildStaysAlive can drive a genuine outbound stdio
// connection hermetically (no external process, no network).
func TestMain(m *testing.M) {
	if os.Getenv("TN_STDIO_MCP") == "1" {
		_ = server.ServeStdio(buildMcpServer([]Tool{mkTool("a"), mkTool("b"), mkTool("c")}, nil, nil))
		return
	}
	os.Exit(m.Run())
}

// TestStdioOutboundChildStaysAlive is the regression guard for the v0.9.0 stdio bug: the child
// process must NOT be killed after connect. It was, when the transport Start ran on a
// timeout/cancel context (mark3labs spawns via exec.CommandContext), so tools/list hit a dead
// pipe. This drives a real outbound stdio server and asserts the tools load AND one executes.
func TestStdioOutboundChildStaysAlive(t *testing.T) {
	cfg := map[string]any{"srv": map[string]any{
		"command":     []any{os.Args[0]},
		"environment": map[string]any{"TN_STDIO_MCP": "1"},
	}}
	src, err := LoadMcp(cfg)
	if err != nil {
		t.Fatalf("load: %v", err)
	}
	defer src.Close()
	if src.Status["srv"] != StatusConnected {
		t.Fatalf("stdio server not connected: status=%v", src.Status["srv"])
	}
	names := rcNames(src.Tools)
	if !names["srv_a"] || !names["srv_b"] || !names["srv_c"] {
		t.Fatalf("stdio tools/list returned %v (child likely killed after connect)", names)
	}
	var srvA Tool
	for _, tl := range src.Tools {
		if tl.Name == "srv_a" {
			srvA = tl
		}
	}
	res, err := srvA.Execute(map[string]any{}, nil)
	if err != nil || res.IsError {
		t.Fatalf("execute srv_a on live stdio child: err=%v res=%+v", err, res)
	}
}

// captureLLM returns an httptest server that records the last decoded request body.
func captureLLM(t *testing.T) (*httptest.Server, func() map[string]any) {
	t.Helper()
	var mu sync.Mutex
	var last map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		mu.Lock()
		last = map[string]any{}
		_ = json.Unmarshal(raw, &last)
		mu.Unlock()
		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"choices":[{"message":{"content":"ok"}}],"content":[{"type":"text","text":"ok"}],"usage":{"prompt_tokens":1,"completion_tokens":1,"total_tokens":2,"input_tokens":1,"output_tokens":1}}`))
	}))
	return srv, func() map[string]any { mu.Lock(); defer mu.Unlock(); return last }
}

func mkTool(name string) Tool {
	return Tool{
		Name:        name,
		Description: name,
		InputSchema: JSONSchema{"type": "object", "properties": map[string]any{}},
		Source:      SourceNative,
		Execute:     func(map[string]any, *ToolContext) (ToolResult, error) { return ToolResult{Output: name}, nil },
	}
}

func TestGap1RequestParamsMergeAndCallerWins(t *testing.T) {
	srv, body := captureLLM(t)
	defer srv.Close()
	tk, _ := CreateToolkit(context.Background(), Options{Builtins: false})

	oa := CreateClient(ClientOptions{BaseURL: srv.URL, Style: StyleOpenAI, Model: "m", APIKey: "k",
		RequestParams: map[string]any{"temperature": 0.42, "chat_template_kwargs": map[string]any{"enable_thinking": false}}})
	_, _ = oa.Run(context.Background(), "hi", tk)
	if body()["temperature"] != 0.42 {
		t.Fatalf("openai temperature = %v", body()["temperature"])
	}
	if body()["chat_template_kwargs"] == nil {
		t.Fatalf("chat_template_kwargs missing")
	}

	an := CreateClient(ClientOptions{BaseURL: srv.URL, Style: StyleAnthropic, Model: "m", APIKey: "k",
		RequestParams: map[string]any{"max_tokens": float64(999)}})
	_, _ = an.Run(context.Background(), "hi", tk)
	if body()["max_tokens"] != float64(999) {
		t.Fatalf("anthropic max_tokens = %v, want 999 (caller wins over 4096)", body()["max_tokens"])
	}
}

func TestGap1BodyTransformAndForbiddenKeys(t *testing.T) {
	srv, body := captureLLM(t)
	defer srv.Close()
	tk, _ := CreateToolkit(context.Background(), Options{Builtins: false})
	c := CreateClient(ClientOptions{
		BaseURL: srv.URL, Style: StyleOpenAI, Model: "m", APIKey: "k",
		RequestParams: map[string]any{"temperature": 0.5, "messages": []any{"nope"}, "tools": []any{1}, "stream": true},
		BodyTransform: func(b map[string]any) map[string]any { b["injected"] = "yes"; delete(b, "temperature"); return b },
	})
	_, _ = c.Run(context.Background(), "hi", tk)
	if body()["injected"] != "yes" {
		t.Fatalf("bodyTransform output not sent")
	}
	if _, ok := body()["temperature"]; ok {
		t.Fatalf("bodyTransform ran before marshal but key survived")
	}
	if _, ok := body()["stream"]; ok {
		t.Fatalf("forbidden 'stream' override leaked into a non-stream call")
	}
	msgs, _ := body()["messages"].([]any)
	if len(msgs) == 0 {
		t.Fatalf("forbidden 'messages' override replaced the real messages")
	}
}

type recordRT struct {
	inner http.RoundTripper
	hits  int
}

func (r *recordRT) RoundTrip(req *http.Request) (*http.Response, error) {
	r.hits++
	return r.inner.RoundTrip(req)
}

func TestGap2HTTPClientInjected(t *testing.T) {
	srv, _ := captureLLM(t)
	defer srv.Close()
	rt := &recordRT{inner: http.DefaultTransport}
	tk, _ := CreateToolkit(context.Background(), Options{Builtins: false})
	c := CreateClient(ClientOptions{BaseURL: srv.URL, Style: StyleOpenAI, Model: "m", APIKey: "k", HTTPClient: &http.Client{Transport: rt}})
	if _, err := c.Run(context.Background(), "hi", tk); err != nil {
		t.Fatalf("run: %v", err)
	}
	if rt.hits == 0 {
		t.Fatalf("injected HTTPClient was not used")
	}
}

func TestGap5OmitEmptyTools(t *testing.T) {
	srv, body := captureLLM(t)
	defer srv.Close()

	empty, _ := CreateToolkit(context.Background(), Options{Builtins: false})
	c1 := CreateClient(ClientOptions{BaseURL: srv.URL, Style: StyleOpenAI, Model: "m", APIKey: "k"})
	_, _ = c1.Run(context.Background(), "hi", empty)
	if _, ok := body()["tools"]; ok {
		t.Fatalf("empty toolkit still sent tools")
	}
	if _, ok := body()["tool_choice"]; ok {
		t.Fatalf("empty toolkit still sent tool_choice")
	}

	one, _ := CreateToolkit(context.Background(), Options{Builtins: false, ExtraTools: []Tool{mkTool("t")}})
	c2 := CreateClient(ClientOptions{BaseURL: srv.URL, Style: StyleOpenAI, Model: "m", APIKey: "k"})
	_, _ = c2.Run(context.Background(), "hi", one)
	if _, ok := body()["tools"]; !ok {
		t.Fatalf("non-empty toolkit dropped tools")
	}
	if body()["tool_choice"] != "auto" {
		t.Fatalf("non-empty toolkit dropped tool_choice")
	}
}

func TestGap4ConversationStoreAccessor(t *testing.T) {
	custom := NewInMemoryConversationStore()
	c1 := CreateClient(ClientOptions{BaseURL: "http://x", Style: StyleOpenAI, Model: "m", Store: custom})
	if c1.ConversationStore() != custom {
		t.Fatalf("accessor did not return the supplied store")
	}

	srv, _ := captureLLM(t)
	defer srv.Close()
	tk, _ := CreateToolkit(context.Background(), Options{Builtins: false})
	c2 := CreateClient(ClientOptions{BaseURL: srv.URL, Style: StyleOpenAI, Model: "m", APIKey: "k"})
	if c2.ConversationStore() == nil {
		t.Fatalf("default store is nil")
	}
	_, _ = c2.Ask(context.Background(), "hi", tk, "conv1")
	hist, _ := c2.ConversationStore().Get("conv1")
	if len(hist) < 2 {
		t.Fatalf("default store did not hold the saved transcript, got %d msgs", len(hist))
	}
}

// startMCP builds a real streamable-HTTP MCP server exposing tools a,b,c.
func startMCP(t *testing.T) (url string, closeFn func()) {
	t.Helper()
	srv := buildMcpServer([]Tool{mkTool("a"), mkTool("b"), mkTool("c")}, nil, nil)
	ts := httptest.NewServer(server.NewStreamableHTTPServer(srv, server.WithStateLess(true)))
	return ts.URL + "/mcp", ts.Close
}

func TestGap7PerServerToolAllowlist(t *testing.T) {
	url, done := startMCP(t)
	defer done()

	allow, err := LoadMcp(map[string]any{"srv": map[string]any{"url": url, "tools": map[string]any{"a": true, "b": true}}})
	if err != nil {
		t.Fatalf("load: %v", err)
	}
	defer allow.Close()
	names := rcNames(allow.Tools)
	if len(names) != 2 || !names["srv_a"] || !names["srv_b"] {
		t.Fatalf("allowlist tools = %v, want srv_a,srv_b", names)
	}

	drop, _ := LoadMcp(map[string]any{"srv": map[string]any{"url": url, "tools": map[string]any{"c": false}}})
	defer drop.Close()
	if n := rcNames(drop.Tools); len(n) != 2 || n["srv_c"] {
		t.Fatalf("drop-list tools = %v, want srv_a,srv_b", n)
	}

	all, _ := LoadMcp(map[string]any{"srv": map[string]any{"url": url}})
	defer all.Close()
	if n := rcNames(all.Tools); len(n) != 3 {
		t.Fatalf("nil filter tools = %v, want all 3", n)
	}
}

func TestGap6ListMcpTools(t *testing.T) {
	url, done := startMCP(t)
	defer done()
	inv, err := ListMcpTools(context.Background(), map[string]any{
		"good": map[string]any{"url": url, "tools": map[string]any{"a": true}}, // filter IGNORED by inventory
		"bad":  map[string]any{"url": "http://127.0.0.1:1/mcp", "timeout": 500},
	})
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(inv.Tools["good"]) != 3 {
		t.Fatalf("inventory should be unfiltered (a,b,c), got %d", len(inv.Tools["good"]))
	}
	if inv.Status["good"] != StatusConnected {
		t.Fatalf("good status = %v", inv.Status["good"])
	}
	if inv.Status["bad"] != StatusFailed {
		t.Fatalf("bad status = %v, want failed", inv.Status["bad"])
	}
}

func TestGap3HangingBoundedAndCancel(t *testing.T) {
	// A server that accepts but never responds → connect/list bounded by timeout.
	// The handler blocks on `unblock` (closed at cleanup, before Close, so Close
	// doesn't wait forever on an in-flight request).
	unblock := make(chan struct{})
	hang := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { <-unblock }))
	defer hang.Close()
	defer close(unblock)

	t0 := time.Now()
	src, err := LoadMcp(map[string]any{"hang": map[string]any{"url": hang.URL + "/mcp", "timeout": 500}})
	if err != nil {
		t.Fatalf("load returned error (per-server timeout should isolate): %v", err)
	}
	defer src.Close()
	if src.Status["hang"] != StatusFailed {
		t.Fatalf("hanging server status = %v, want failed", src.Status["hang"])
	}
	if dt := time.Since(t0); dt > 5*time.Second {
		t.Fatalf("not bounded by timeout: took %v", dt)
	}

	// Parent-ctx cancellation aborts the whole build with ctx.Err().
	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	if _, err := LoadMcpWithContext(ctx, map[string]any{"hang": map[string]any{"url": hang.URL + "/mcp", "timeout": 60000}}); err == nil {
		t.Fatalf("cancelled ctx should error the whole load")
	}
}

func TestDisableToolsAndSkills(t *testing.T) {
	url, done := startMCP(t)
	defer done()
	tk, err := CreateToolkit(context.Background(), Options{
		Builtins:     false,
		McpConfig:    map[string]any{"srv": map[string]any{"url": url}},
		DisableTools: []string{"srv_b"},
	})
	if err != nil {
		t.Fatalf("toolkit: %v", err)
	}
	if _, ok := tk.Get("srv_b"); ok {
		t.Fatalf("DisableTools did not drop srv_b")
	}
	if _, ok := tk.Get("srv_a"); !ok {
		t.Fatalf("DisableTools dropped too much (srv_a gone)")
	}

	// DisableSkills drops a skill by name from the catalog.
	sk, err := CreateToolkit(context.Background(), Options{
		Builtins:      false,
		Skills:        []SkillDef{{Name: "keep", Description: "k", Content: "k"}, {Name: "drop", Description: "d", Content: "d"}},
		DisableSkills: []string{"drop"},
	})
	if err != nil {
		t.Fatalf("skills toolkit: %v", err)
	}
	prompt := sk.SkillsPrompt()
	if !strings.Contains(prompt, "keep") || strings.Contains(prompt, "**drop**") {
		t.Fatalf("DisableSkills wrong; prompt:\n%s", prompt)
	}
}

func rcNames(ts []Tool) map[string]bool {
	m := map[string]bool{}
	for _, t := range ts {
		m[t.Name] = true
	}
	return m
}
